### PR TITLE
Adjust mainnet parameters for litecoin

### PR DESCRIPTION
**Pull request by ysangkok**
I didn't include any checkpoints or all the other dns seeds.

### CODE REVIEW

**Comment by Melvillian** [comment]
It would be nice to get this landed, because currently ltcd will try to sync with bitcoin peers, and throw endless errors about the proof of work limit being too high. I've run ltcd with this PR and it syncs correctly for me.

**Comment by ysangkok** [comment]
There are minor issues with this PR though. For example, I don't know what to put as BIP activation heights: https://github.com/ltcsuite/ltcd/pull/2/commits/327be039a288e7ed6371427cdba14696e903d6c2#diff-3203a6040cfe5ff98d1f499872880defR238

**Comment by Melvillian** [comment]
Do those BIPs exist in Litecoin? This Issue suggests they are in litecoin

https://github.com/litecoin-project/litecoin/issues/244

If they're not in Litecoin, then you can set the activation heights to be max of the int32 type

**Comment by ysangkok** [comment]
I can't figure that out, but if I tried setting them to -1, and that won't work.

BTW, @Melvillian  are you sure you synced to May blocks? My ltcd seems to reject block 5934f52b03a9bc5c86156533df422bb17ffef0873ac407106cb03ffdad94d1ec from april because it has got an empty witness stack. But hasn't activated yet, so I think I got something wrong with either activation timestamps, or voting thresholds. Dunno what yet. Very weird if it works for you.

**Comment by losh11** [comment]
https://github.com/litecoin-project/litecoin/blob/master/src/chainparams.cpp#L90

Line 90-98 concerns BIP9 activation heights for CSV and SW. SW isn't active on mainnet yet, that's probably why setting activation heights to -1 isn't working. You can try connecting your node only to Litecoin Core 0.13.2.1 and using the console to debug issues to do with syncing.

**Comment by ysangkok** [comment]
@losh11 I am pretty sure those are not heights, but UNIX time stamps of the voting period (STARTED from [BIP9](https://github.com/bitcoin/bips/blob/master/bip-0009.mediawiki#state-transitions)). -1 would mean segwit has been active since the genesis block, which makes validation fail for all blocks. I just don't understand why that block from april is supposed to have a witness stack. The parameters obviously work in Litecoin Core, so why do they not work in ltcd?

**Comment by Melvillian** [comment]
@ysangkok I haven't synced all the way to chainhead yet, but I meant that at least it had begun syncing correctly when I added your PR.

**Comment by OttoAllmendinger** [comment]
👍 just for the fixes in `cmd/.../config.go`, refuses to compile otherwise

File: chaincfg/params.go [hunk 1]
```diff
@@ -225,61 +225,37 @@ type Params struct {
...
-		{352940, newHashFromStr("000000000000000010755df42dba556bb72be6a32f3ce0b6941ce4430152c9ff")},
-		{382320, newHashFromStr("00000000000000000a8dc6ed5b133d0eb2fd6af56203e4159789b092defd8ab2")},
 	},
 
 	// Consensus rule change deployments.
 	//
 	// The miner confirmation window is defined as:
 	//   target proof of work timespan / target proof of work spacing
-	RuleChangeActivationThreshold: 1916, // 95% of MinerConfirmationWindow
+	RuleChangeActivationThreshold: 1512, // 75% of MinerConfirmationWindow
```
Comment (OttoAllmendinger) [inline chaincfg/params.go]: This looks off: https://github.com/litecoin-project/litecoin/blob/master/src/chainparams.cpp#L85

I had the same problem with the false Segwit block in april

Testing now with 6048/8064 parameters

Comment (OttoAllmendinger) [reply chaincfg/params.go]: looks good with changed parameters!

```
14:31:26 2017-05-10 [INF] BMGR: Processed 1 block in the last 1m59.66s (105 transactions, height 1201447, 2017-05-10 14:29:44 +0200 CEST)
14:32:15 2017-05-10 [INF] BMGR: Processed 1 block in the last 48.48s (53 transactions, height 1201448, 2017-05-10 14:31:24 +0200 CEST)
```

however I had to comment out the threshold cache initialization, I think it did not detect the parameter change correctly: https://github.com/ltcsuite/ltcd/blob/master/blockchain/chain.go#L1426-L1428

Comment (ysangkok) [reply chaincfg/params.go]: works for me too :+1:  thanks

**Comment by ysangkok** [comment]
Seems like there might be another problem: 

```
0:53:30 2017-05-11 [INF] BMGR: Rejected block 
852004886a4f338aa773cd009cd4012f928d7fd8c8ad3e3659ba5cf2b2431cb6
from 155.41.120.249:9333 (outbound): failed to validate input 
ce385e55fb2a73fa438426145b074f08314812fa3396472dc572b3079e26e0f9:0 
which references output 
6f45775d5f5b62bdd1527151ddb7f6b19a01bbcb6c3ecebf4c581f11851c1e37:0
 - element size 3556 exceeds max allowed size 520 (input witness [[77 245 1 ... ]],
input script bytes 
22002091fb57fbeb6404cb412c104fcfb6bbcbe144884135d260879738f34f97d3d0ed, 
prev output script bytes a9140e5da99fdabb8f78741b6d7722b8e2647d7b5ec887)

**Comment by JimmyMow** [comment]
+1 

I was getting that output as well earlier

**Comment by benma** [comment]
This is a segwit transaction. The scriptSig is a BIP16 redeemScript containing the witness program

The length of the witness script is not validated correctly, see:


This commit should fix it, I didn't verify it yet though (feel free to integrate it into this PR):

https://github.com/benma/ltcd/commit/9387d45795073f5c39d6e6772135def03c4933c9

**Comment by benma** [comment]
This is a P2WSH segwit transaction. The scriptSig is a BIP16 redeemScript containing the witness program

The length of the witness script is not validated correctly, see:
https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#witness-program

(the code enforces a 520 limit also to the witnessScript, where the limit is 10,000).

This commit should fix it, I didn't verify it yet though (feel free to integrate it into this PR):
https://github.com/benma/ltcd/commit/da3c8c6dc6c304bcc17a6179fc9584056df8acd4

**Comment by davecgh** [comment]
I'll comment here too for reference.  `btcd` master doesn't yet support segwit, so there are certainly segwit-specific things that aren't in master.

I would like to point out that reading that particular error message, it seems pretty clear that it's applying the `MaxScriptElementSize` to the overall witness script data push instead of `MaxScriptSize` as required by the segwit specification.  The `MaxScriptElementSize` still applies to each individual push within the witness script, but not to the overall witness script itself.

**Comment by benma** [comment]
@davecgh thx. I tried to apply this in the commit referenced above. Is it correct that the check for `MaxScriptElementSize` can be skipped altogether in this function because it will be checked in `executeOpcode()` anyway?

If not, I will re-add the check.

**Comment by davecgh** [comment]
@benma: I have not yet gone through the segwit code with a fine tooth comb, but I'm almost positive removing the check from the witness program verify would lead to a consensus break.  The reason for that is because you could possibly construct a script that has an push that exceeds the maximum 520 bytes that is never actually executed in the witness program such as would be the case in an `OP_IF` with a 0 on the stack.

My _guess_ is that the witness script is supposed to be popped from the stack before checking the rest of the stack items just the same way that p2sh works.  However, that would need to be verified by looking at the litecoin C++ code for verifying witness programs.

**Comment by benma** [comment]
Thanks a lot.

> My guess is that the witness script is supposed to be popped from the stack before checking the rest of the stack items just the same way that p2sh works. 

A previous version of my commit did this. I will restore it shortly.

Edit: done, link to commit above is updated.

**Comment by ysangkok** [comment]
@benma  are you sure that's right? It's losing peers all the time, for me.

```
09:51:30 2017-05-11 [INF] BMGR: Syncing to block height 1201889 from peer 47.93.232.59:9333
09:51:30 2017-05-11 [INF] BMGR: New valid peer 88.198.230.142:9333 (outbound) (/Satoshi:0.13.2/)
09:52:13 2017-05-11 [INF] BMGR: Lost peer 47.93.232.59:9333 (outbound)
09:52:13 2017-05-11 [INF] BMGR: Syncing to block height 1201889 from peer 88.198.230.142:9333
09:52:43 2017-05-11 [INF] PEER: Disconnecting non-segwit peer 128.199.177.182:9333 (outbound), isn't segwit segwit enabled and we need more segwit enabled peers
09:52:43 2017-05-11 [INF] BMGR: New valid peer 128.199.177.182:9333 (outbound) (/Satoshi:0.10.4/)
09:52:43 2017-05-11 [INF] BMGR: Lost peer 128.199.177.182:9333 (outbound)
09:52:43 2017-05-11 [INF] PEER: Disconnecting non-segwit peer 93.189.145.169:9333 (outbound), isn't segwit segwit enabled and we need more segwit enabled peers
09:52:43 2017-05-11 [INF] BMGR: New valid peer 93.189.145.169:9333 (outbound) (/Satoshi:0.8.6.2/)
09:52:43 2017-05-11 [INF] BMGR: Lost peer 93.189.145.169:9333 (outbound)
09:52:45 2017-05-11 [INF] BMGR: Lost peer 88.198.230.142:9333 (outbound)
```

I can understand why its dropping non-segwit peers, but the other peers are also getting lost within a very short time.

But it does sync to a current height. So thanks for the fix.

**Comment by benma** [comment]
@ysangkok not sure, no. I am still in the process of syncing, no where near the problematic block/transaction.

@davecgh 

Grepping for `MaxScriptElementSize`, it seems the only place it is used to check the element size is in `executeOpcode()`, and the code in there checks the size of the element before it checks if it should be ignored due to conditionals, etc. So maybe the check in verifyWitnessProgram is redundant after all?

**Comment by OttoAllmendinger** [comment]
It syncs correctly for me with https://github.com/ltcsuite/ltcd/pull/2/commits/baf1c2992e5f304038831efaca083611f196bada

Good job @benma !

**Comment by ghost** [comment]
Dear all,
it also syncs well with ltcsuite/ltcd@baf1c29 for me.
Thank you for the work.

**Comment by ysangkok** [comment]
there is a bug in benma's commit, roasbeef knows about it and will push a fix after segwit get merged into btcd mainline

**Comment by benma** [comment]
@ysangkok can you elaborate?

**Comment by ysangkok** [comment]
@benma , no, I am just parroting. You'd have to ask roasbeef.

File: chaincfg/params.go [hunk 2]
```diff
@@ -289,29 +265,29 @@ var MainNetParams = Params{
...
 		},
 	},
 
 	// Mempool parameters
 	RelayNonStdTxs: false,
 
 	// Address encoding magics
-	PubKeyHashAddrID:        0x00, // starts with 1
+	PubKeyHashAddrID:        48, // starts with L
 	ScriptHashAddrID:        0x05, // starts with 3
```
Comment (losh11) [inline chaincfg/params.go]: From the 0.13.3 release, the P2SH prefix is `M` by default.

**Pull request closed without merging**